formats/json/commonMain/src/kotlinx/serialization/json/JsonElement.kt

```diff
@@ -61,6 +63,17 @@ public fun JsonPrimitive(value: Number?): JsonPrimitive {
     return JsonLiteral(value, isString = false)
 }
 
+/**
+ * Creates [JsonPrimitive] from the given number.
+ */
+public fun JsonLiteralNumber(value: String?): JsonPrimitive {
+    if (value == null) return JsonNull
+    if (!jsonLiteralNumberRegex.matches(value)) throw JsonEncodingException("not a number: $value")
```

Comment:
Regex validating would block various other use-cases, like the one mentioned here: https://github.com/Kotlin/kotlinx.serialization/issues/1405#issuecomment-813381394

Now I'm a bit hesitant about whether we should support this (because we can't parse it back as the same raw value, but it is not needed), but it looks like a good use case. Supporting it would also require renaming this function into something like `JsonUnquotedLiteral` to avoid any connection to numbers, so it's better to think about this from the beginning.

Reply:
In order to keep this PR focused I just wanted to completely prevent encoding anything except JSON numbers. I thought that allowing encoding raw strings would be a much larger discussion that would prevent an incremental improvement. Of course, changing this PR to allow for encoding raw strings is also possible!